### ISSUE TITLE
A master table and other suggestions

### ISSUE THREAD

**Issue by sayakpaul** [AdityaKane2001/regnety#15]
@AdityaKane2001 

You have done an extremely complete and great job so far! Here's what I would suggest going further to make things a bit more manageable:

* Create a master table in the README with the following columns (you can change the names, of course): MF, Setup (keep the entires brief), Validation Top-1 Acc, Model Location. 
* Try working on a script to measure how many images per second a given RegNet is able to process on a commodity GPU like K80.

Cc: @MorganR

**Comment by AdityaKane2001** [comment]
@sayakpaul 

I'll make the following scripts:

- predict.py : for predictions
- evaluation.py: for evaluating the models on user data
- benchmark.py: for speed and performance measurement of individual components (input pipeline, model)

And the master table as you have mentioned.

**Comment by sayakpaul** [comment]
Input pipeline may be discarded. But for predictions and user data, I'd suggest Colab Notebooks. Think it from a developer's perspective. Won't it be much easier to simply spin up a Colab Notebook where one can select a particular RegNet model, run predictions, and evaluate? Also, if you mean running evaluation on the ImageNet-1k validation split, then a script would be preferred. 

For benchmarking (i.e. how many images/second a particular RegNet model can process), we can either use a script or a Colab Notebook.

**Comment by AdityaKane2001** [comment]
Got it. Will use Colab for predict and evaluation, script for benchmarking.

**Comment by AdityaKane2001** [comment]
@sayakpaul 

I am planning to keep the best checkpoints in a public GCS bucket as well as upload them to GH. I would've loved to also provide h5 versions of the models on GH,  but it requires a bit of investigation and given the time frame I'd like to keep that as a stretch goal. 

Apart from that, I'm working on the notebook and I'll share it with you by today evening. Same goes for the benchmarking script, I'll create a PR by today evening.

**Comment by AdityaKane2001** [comment]
> I am planning to keep the best checkpoints in a public GCS bucket as well as upload them to GH. 

About this, I realized that it is redundant to do that, since we're going to upload these to TFHub.
So finally, only loading from TF Hub, prediction and finetuning in the notebook.

**Comment by AdityaKane2001** [comment]
@sayakpaul 

I have completed a first draft of the notebook. Please take a look at it [here](https://colab.research.google.com/drive/1NXIWVJSqRxKgJzhUGfaod-HLMxPqCBgG?usp=sharing). Please share your thoughts on the same.

I will start working parallelly on the master table and the benchmarking script.

**Comment by sayakpaul** [comment]
The notebook looks fantastic.

**Comment by AdityaKane2001** [comment]
@sayakpaul @MorganR 

> Try working on a script to measure how many images per second a given RegNet is able to process on a commodity GPU like K80.

For the benchmarking script, I was thinking of keeping it very simple:
- initialize model
- warm up device
- warm up dataset iterator
- measure time using timeit/ time.time()

I wanted to just `timeit` the `.evaluate()` call and divide it by number of images. I understand that this is naive, but frankly I don't know of any other simple way to achieve this. Could you please tell a bit about this?

**Comment by sayakpaul** [comment]
> I wanted to just timeit the .evaluate()

Why would you want to call `evaluate()`. Shouldn't this be `.predict()`? Also, be sure to NOT put any external data preprocessing code outside of the timing calculation. 

A few resources that you may find useful:
* https://docs.fritz.ai/training/python-sdk/benchmark/
* https://github.com/cgnorthcutt/benchmarking-keras-pytorch
* https://github.com/tensorflow/benchmarks

@MorganR may have additional thoughts on this.

**Comment by AdityaKane2001** [comment]
> Shouldn't this be .predict()?

Sorry, it should be predict(). My bad. 

> https://docs.fritz.ai/training/python-sdk/benchmark/

If it works properly I will use this.

**Comment by AdityaKane2001** [comment]
In the following week, these are the things I need to complete (which I am aware of, arranged in order of priority):

- TFHub PR documentation.
- Benchmarking script.
- Merge current dev branch, it has some changes made over the past few days.
- Repository documentation.
- (Stretch goal) Convert models to TFLite.

Please tell me if I'm missing something. Will add this to the project view.

**Comment by sayakpaul** [comment]
Shouldn't there be these two as well?

* Colab Notebook finalization? 
* Relook the implementation for v2 if needed?

**Comment by AdityaKane2001** [comment]
> Colab Notebook finalization?

I'll add that, yes.

> Relook the implementation for v2 if needed?

I am going over it right now. Will post here if I find something.

**Comment by AdityaKane2001** [comment]
> https://docs.fritz.ai/training/python-sdk/benchmark/

This isn't working as I am not able to sign in to the app.

Since we're only aiming to measure the inference speed, I think using a minimal will be sufficient.

**Comment by sayakpaul** [comment]
Yup. Agreed.

**Comment by AdityaKane2001** [comment]
@sayakpaul @MorganR 

Pushed [notebook](https://github.com/AdityaKane2001/regnety/blob/dev/RegNetY_models_in_TF_2_5.ipynb).

Benchmarking script [link](https://github.com/AdityaKane2001/regnety/blob/dev/regnety/benchmark.py) for reference.

**Comment by AdityaKane2001** [comment]
@sayakpaul @MorganR 

I was able to spin up a K80 instance and here are the inference times.

K80 200MF Inference of 262144 images took 399.22830510139465 seconds. images/second = 656.6267888581235
200MF params: 3,231,775

K80 400MF Inference of 262144 images took 604.1937155723572 seconds. images/second = 433.87409243684215
400MF params: 4,058,881

K80 600MF Inference of 65536 images took 182.1467649936676 seconds. images/second = 359.7977707826893
600MF params: 6,214,479

K80 800MF Inference of 65536 images took 213.98049759864807 seconds. images/second = 306.2709019535155
800MF params 6,501,119

**Comment by MorganR** [comment]
Great work! The benchmarking script looks good too 👍

**Comment by sayakpaul** [comment]
This one could be closed as well?

**Comment by AdityaKane2001** [comment]
If it's okay, I'll close #10 instead. We can keep this one for communication.

**Comment by AdityaKane2001** [comment]
@sayakpaul 

The grouped conv issue is resolved. Our model works on GPUs and TPUs.  On CPU, the model cannot be trained at all, but we can run inference with any batch size. (🥳)

We have a small issue related to loading of checkpoints. The checkpoints are not getting loaded with the error message:

```
Two checkpoint references resolved to different objects (<regnety.models.blocks.Stem object at 0x7f9aa186bf50> and <regnety.models.blocks.PreStem object at 0x7f9aa185fc90>).
23-Aug-21 11:07:52 WARNING : Inconsistent references when loading the checkpoint into this object graph. Either the Trackable object references in the Python program have changed in an incompatible way, or the checkpoint was generated in an incompatible program.

(and many similar error messages)
```
Do you happen to know why this may be happening?

Gist [here](https://colab.research.google.com/gist/AdityaKane2001/424e9f482c52b2e205d344d6cc27a8d9/groupedconvissue.ipynb?authuser=5).


/cc @MorganR

**Comment by sayakpaul** [comment]
This is great news, Aditya. Thanks for letting us know. Since the checkpoints contain custom layers, maybe specifying their configurations while loading them would be helpful. More [here](https://keras.io/api/models/model_saving_apis/#getconfig-method).

This is why I prefer `.h5` weights. Just instantiate your model class and once that is done call `load_weights()` on it.

**Comment by AdityaKane2001** [comment]
Actually, I've implemented `get_config` methods for all the custom layers. Thus, that shouldn't be an issue. I believe this bypasses the requirement for passing `custom_objects` argument while loading from SavedModel. [reference](https://www.tensorflow.org/tutorials/keras/save_and_load#saving_custom_objects)

> This is why I prefer .h5 weights. Just instantiate your model class and once that is done call load_weights() on it.

I see. I'll try to make to those files from the SavedModel format. If that works, then I'll upload the h5s instead.

On a related note, I don't think subclassing `tf.keras.Model` for the layers is a good idea as we have trained using models that subclass Layer. I'll still do that on a separate branch for debugging.

**Comment by sayakpaul** [comment]
> On a related note, I don't think subclassing tf.keras.Model for the layers is a good idea as we have trained using models that subclass Layer. I'll still do that on a separate branch for debugging.

Going forward subclassed layers for implementing block modules might not be a good idea. I think I provided a couple of examples to you via GChat on that. Right?

**Comment by AdityaKane2001** [comment]
> But that said, your final model is anyway a tf.keras.Model. However, I am not sure why did you bring up it here.

Yes, the final model is a Sequential. I brought this up as an FYI as it won't be possible to load the existing trained weights in the new models. So we'd need to retrain those models.

I think the takeaway from this discussion is that we should stick to SavedModel wherever possible. It completely avoids any further complications, and we can load and play with the model without requiring the source code, which is a requirement for h5. 

I'll create an issue in the TF repo for the checkpoint loading issue.
